scripts: west_commands: build: support --cmake-opt

thorsten-klein opened the pull request:
## Why?
When using west aliases, it is helpful for users to give cmake arguments, e.g. to make building specific variants as easy as possible. Currently `--` must be used, e.g.
```
[alias]
build-llvm = build -- -DZEPHYR_TOOLCHAIN_VARIANT=llvm
build-codechecker = build -- -DZEPHYR_SCA_VARIANT=codechecker
build-llvm-codechecker = build -- -DZEPHYR_TOOLCHAIN_VARIANT=llvm -DZEPHYR_SCA_VARIANT=codechecker
```
With this approach it is not possible anymore to specify west arguments when the alias is used (e.g. board `-b native_sim`), as the argument comes behind `--` and therefore it is handled as cmake argument.

## Solution
Support for `--cmake-opt` argument is added which allows specifiying cmake options via `west build --cmake-opt=<arg>`, in order to avoid `--` in alias commands. This argument can be specified multiple times, e.g.
```
[alias]
build-llvm = build --cmake-opt=-DZEPHYR_TOOLCHAIN_VARIANT=llvm
build-codechecker = build --cmake-opt=-DZEPHYR_SCA_VARIANT=codechecker
build-llvm-codechecker = build --cmake-opt=-DZEPHYR_TOOLCHAIN_VARIANT=llvm --cmake-opt=-DZEPHYR_SCA_VARIANT=codechecker

# or even easier: build-llvm can be reused
build-llvm-codechecker = build-llvm --cmake-opt=-DZEPHYR_SCA_VARIANT=codechecker
```

thorsten-klein committed:
scripts: west_commands: build: support cmake arguments via --cmake-opt

Support cmake options specified via argument --cmake-opt, in order to
avoid `--` in alias commands.

Signed-off-by: Thorsten Klein <thorsten.klein@bshg.com>

pdgendt commented on scripts/west_commands/build.py:
```diff
@@ -119,6 +119,10 @@ def do_add_parser(self, parser_adder):
+        group.add_argument('--cmake-opt', action='append',
+                           help='''same as '-- cmake_opt', but it prevents
+                           using '--' (e.g. for usage in alias commands);
+                           may be given multiple times.''')
```
Don't use both `cmake_opt` and `cmake_opts`, it is confusing. Set a default instead and extend in `_parse_remainder`

```suggestion
        group.add_argument('--cmake-opt', action='append',
                           dest="cmake_opts", default=[],
                           help='''same as '-- cmake_opt', but it prevents
                           using '--' (e.g. for usage in alias commands);
                           may be given multiple times.''')
```

(There's a debug print for `cmake_opts`)

pdgendt commented on scripts/west_commands/build.py:
```diff
@@ -119,6 +119,11 @@ def do_add_parser(self, parser_adder):
         group.add_argument('-c', '--cmake', action='store_true',
                            help='force a cmake run')
+        group.add_argument('--cmake-opt', action='append',
+                           dest="cmake_opts", default=[],
```
```suggestion
                           dest="cmake_opts",
```

pdgendt commented on scripts/west_commands/build.py:
```diff
@@ -305,7 +310,8 @@ def _parse_remainder(self, remainder):
                 remainder = remainder[1:]
             if remainder:
-                self.args.cmake_opts = remainder
+                self.args.cmake_opts = self.args.cmake_opts or []
```
```suggestion
```

pdgendt commented:
> Or, you could just use shell functions like this one:
> 
> ```shell
> wb_llvm() { west build "$@" -- -DZEPHYR_TOOLCHAIN_VARIANT=llvm; }
> wb_llvm -b native_sim ...
> ```
> 
> This is exactly the type of problems shells are designed to solve. They can solve this one simply and can do a lot more.
> 
> So this looks like "feature creep" and unnecessary maintenance and complexity to me.

I have to disagree here, adding shell aliases/functions scatters the commands to different places in the OS/workspace. I think this is a useful feature, and adds little maintenance overhead.

pdgendt commented on scripts/west_commands/build.py:
```diff
@@ -407,10 +412,7 @@ def _parse_test_item(self, test_item, board):
-                        self.args.cmake_opts.extend(args)
-                    else:
-                        self.args.cmake_opts = args
+                    self.args.cmake_opts.extend(args)
```
Can't `self.args.cmake_opts` still be `None` here? (same for change further down the function)

thorsten-klein committed:
scripts: west_commands: build: remove shadowed function argument

Remove _run_cmake argument 'cmake_opts' which is never used and shadowed
by a local variable.

Signed-off-by: Thorsten Klein <thorsten.klein@bshg.com>

thorsten-klein committed:
scripts: west_commands: tests: test_build: add tests for west build args

Add tests to ensure 'west build' arguments are correctly forwarded to
CMake in the correct order.

Signed-off-by: Thorsten Klein <thorsten.klein@bshg.com>

tejlmand reviewed (dismissed):
PR is not needed.

> Currently -- must be used,

Help text for `west build` is imprecise on this.

First positional argument to `west build` is treated as the source dir, remaining positional arguments are passed to CMake.

See also: 
https://github.com/zephyrproject-rtos/zephyr/blob/a600317285ff82884055c2135a2ec39af7d42191/scripts/west_commands/build.py#L301-L309

So as long as you use `west build -b <board> <source>` then you don't need `--` and can just do: `west build -b <board> <source> -DFOO=bar`.
Only when omitting source dir, you need `--`, such as in this case:
```
$ cd <app-folder>
$ west build -b <board> -- -DFOO=bar 
```

to avoid using `--` in this case, one can just use a `.`, like this:
```
$ cd <app-folder>
$ west build -b <board> . -DFOO=bar 
```

Other places where source-dir can be omitted is when running an incremental, non-pristine, build, but that to can be handled easily if goal is to allow passing arguments to CMake without the use of `--`.

also see: 
https://github.com/zephyrproject-rtos/west/blob/11302e1ee5af96c9a0cf8c7436992936b1bbd848/src/west/commands.py#L196
https://github.com/zephyrproject-rtos/west/blob/11302e1ee5af96c9a0cf8c7436992936b1bbd848/src/west/commands.py#L214

tejlmand reviewed (approved):
I share the same concerns as @marc-hb here: https://github.com/zephyrproject-rtos/zephyr/pull/101457#issuecomment-3712611876

I do see the usefulness and reason for wanting this in context of `west alias`, but  i'm concerned we can end up having to not allow positional arguments in west extension commands, simply because positional arguments in alias command are never going to be compatible with alias.

I'm not approving this PR because of the wish for having `--cmake-opt` as keyword argument addition to positional argument for the sake of alias.

I'm approving the PR because having `--cmake-opt` in addition aligns it with the build argument `--build-opt`.
Allowing both `--cmake-opt` and `--build-opt` gives a more consistent interface to `west build`. 

Implementation itself looks ok, just a small comment regarding help text.

marc-hb commented on scripts/west_commands/build.py:
```diff
@@ -120,6 +120,13 @@ def do_add_parser(self, parser_adder):
+                           help='''same as using '-- cmake_opt' but avoid the
+                           end-of-options marker '--' (e.g. in alias commands);
+                           those options are passed to cmake first, so they can
+                           be overridden via '-- cmake_opt';
```
@henrikbrixandersen just merged this. I'll submit a small fixup and we can discuss there.

henrikbrixandersen replied:
Sorry, merged just before you commented.

marc-hb replied:
>  I'll submit a small fixup and we can discuss there.

- #102419
